src/emqx_router.erl

```diff
@@ -249,8 +259,19 @@ delete_trie_route(Route = #route{topic = Topic}) ->
 %% @private
 -spec(trans(function(), list(any())) -> ok | {error, term()}).
 trans(Fun, Args) ->
-    case mnesia:transaction(Fun, Args) of
-        {atomic, Ok} -> Ok;
-        {aborted, Reason} -> {error, Reason}
+    mnesia:async_dirty(Fun, Args).
+
+lock_router() ->
+    %% if Retry is not 0, global:set_lock could sleep a random time up to 8s.
+    %% Considering we have a limited number of brokers, it is safe to use sleep 1 ms.
+    case global:set_lock({?MODULE, self()}, [node() | nodes()], 0) of
+        false ->
+            %% Force to sleep 1ms instead.
+            timer:sleep(1),
```

Comment:
should we pass down the caller's pid here and check `process_alive` here ?
in case `false`, we return `{error, ignored}` to the `do_add_route(Topic, Dest)`, caller.

and finally in `emqx_broker` we need to change `subscribe/3` to  
```
  subscribe(Topic, SubId, SubOpts0) when is_binary(Topic), ?is_subid(SubId), is_map(SubOpts0) ->
      SubOpts = maps:merge(?DEFAULT_SUBOPTS, SubOpts0),
      case ets:member(?SUBOPTION, {SubPid = self(), Topic}) of
          false -> %% New
              case do_subscribe(Topic, SubPid, with_subid(SubId, SubOpts)) of
                  {error, ignored} -> ok;
                  _ -> ok = emqx_broker_helper:register_sub(SubPid, SubId)
              end;
          true -> %% Existed
              set_subopts(SubPid, Topic, with_subid(SubId, SubOpts)),
              ok %% ensure to return 'ok'
      end.
```

Reply:
it seems a lot more complicated if we do not call `ok = emqx_broker_helper:register_sub(SubPid, SubId)`

Reply:
I had simlar fix on another branch.

https://github.com/qzhuyan/emqx/commit/e679fc91ce451190d2e0a29a8df5431666660459#diff-25a6634263c1b1f6fc4697a04e2b9904ea4b042a89af59dc93ec1f5d44848a26

from test result, It didn't help that much since the process usually won't die even when clients closed the connection, the 'tcp closed' message is in the msg box of the process while the process is waiting for the subscribe call to return or timeout. I am thinking we should have two processes, one to handle transport and another interact with MQTT layer as the old version.

Anyway this kind of improvement could be addressed in separate branch.